components/suggest/src/db.rs

```diff
@@ -207,7 +209,11 @@ impl<'a> SuggestDao<'a> {
     }
 
     /// Fetches Suggestions of type Amp provider that match the given query
-    pub fn fetch_amp_suggestions(&self, query: &SuggestionQuery) -> Result<Vec<Suggestion>> {
+    pub fn fetch_amp_suggestions(
+        &self,
+        query: &SuggestionQuery,
+        provider: SuggestionProvider,
```

Comment:
YES! +1. Let's [make-invalid-states-unrepresentable](https://geeklaunch.io/blog/make-invalid-states-unrepresentable/)!

Actually, I was gonna suggest adding this exact enum `AmpSuggestionType` to the `Provider` so that consumers have to pick either variant (but not both!) when they query. Then I realized that they could still "violate" that as they could pass both `Mobile` and `Desktop` to `SuggestionQuery::providers` 😂.

Reply:
Oooh, thank you for that link, it's going into my bookmarks for sure! 🏆 I think [this Aphyr article](https://aphyr.com/posts/341-hexing-the-technical-interview) (it's a _read_! :joy:) was the first time I heard that expression:

> “But… those aren’t even the same type. That’s… that’s illegal.”
>
> “If it were meant to be illegal,” you remind him sagely, “Sun Microsystems would have made it unrepresentable.”

I liked your second point, too:

> Then I realized that they could still "violate" that as they could pass both `Mobile` and `Desktop` to `SuggestionQuery::providers`.

Yeah, it's tricky; I think this is the first time that we're segmenting suggestions by platform like this! I'd be curious to see if we end up with similar cases in the future—and if a more fluent way of expressing this separation emerges from those other cases! For now, it feels like the list of providers is the most sensible abstraction we have, but that might change as we add providers and new types of filters.